src/merkle_tree.rs

```diff
@@ -0,0 +1,481 @@
+use sha3::{Digest, Sha3_256};
+
+use crate::direction::Direction;
+use crate::proof_of_inclusion::ProofOfInclusion;
+
+use super::merkle_hash::MerkleHash;
+
+/// A Merkle Tree implementation
+///
+/// # Methods
+/// - `new_from_hashes`: Creates a new MerkleTree from a list of hashes.
+/// - `new_from_hasables`: Creates a new MerkleTree from a list of objects that are hashable.
+/// - `root`: Returns the root of the Merkle Tree, which is the Merkle Root.
+/// - `verify`: Verifies that a given hash is contained in the Merkle Tree.
+/// - `proof_of_inclusion`: Returns a proof of inclusion for a given hash in the Merkle Tree.
+#[derive(Debug)]
+pub struct MerkleTree {
+    levels: Vec<Vec<MerkleHash>>,
+}
+
+impl MerkleTree {
+    /// Creates a new MerkleTree from a list of hashes.
+    pub fn new_from_hashes(hashes: Vec<MerkleHash>) -> MerkleTree {
+        let mut tree = MerkleTree { levels: vec![] };
+        MerkleTree::build_tree(&mut tree, hashes);
+        tree
+    }
+
+    /// Creates a new MerkleTree from a list of objects that can be converted to byte slices (== that are hashable).
+    pub fn new_from_hashables(data: Vec<impl AsRef<[u8]>>) -> MerkleTree {
+        let hashes = data
+            .iter()
+            .map(|d| {
+                let mut hasher = Sha3_256::new();
+                hasher.update(d);
+                let result = hasher.finalize();
+                MerkleTree::bytes_to_hex(&result)
+            })
+            .collect();
+
+        MerkleTree::new_from_hashes(hashes)
+    }
+
+    /// Recursive function that builds the Merkle Tree from a list of hashes.
+    fn build_tree(tree: &mut MerkleTree, mut hashes: Vec<MerkleHash>) {
+        if hashes.len() == 1 {
+            tree.levels.push(hashes.clone());
+            return;
+        }
+
+        if hashes.len() % 2 != 0 {
+            let last = &hashes[hashes.len() - 1];
+            hashes.push(last.clone());
+        }
+
+        tree.levels.push(hashes.clone());
+
+        let mut next_hashes = vec![];
+        for i in (0..hashes.len()).step_by(2) {
+            let left = hashes[i].clone();
+            let right = if i + 1 < hashes.len() {
+                hashes[i + 1].clone()
+            } else {
+                left.clone()
+            };
```

Comment:
```suggestion
            let right = hashes[i+1].clone();
```

I think here `if` is unnecessary cause `i + 1 < hashes.len()` is always true when iterating with step 2 a vector of even length.

_Ex: Vector of length 6 is iterated 3 times: i = 0, 2, 4. 4 + 1 = 5. 5 < 6._

Given that in previous lines of code you forced `hashes.len()` to be even (by pushing it's last element to the vector itself if it’s length was odd) there is no need of doing this.

Let me know if I got it wrong.

Reply:
Thats true thanks!!! I thought the if was necessary but it wasn't. In the meanwhile I was testing the tree and found an error which I also fixed, when the tree was created with only one element it crashed at adding more because of a check that was missing. :))